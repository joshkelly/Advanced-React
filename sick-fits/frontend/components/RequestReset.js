import { gql, useMutation } from '@apollo/client';
import Router from 'next/router';
import useForm from '../lib/useForm';
import DisplayError from './ErrorMessage';
import Form from './styles/Form';
import { CURRENT_USER_QUERY } from './User';

const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    sendUserPasswordResetLink(email: $email) {
      code
      message
    }
  }
`;

export default function RequestReset() {
  const { inputs, handleChange, resetForm } = useForm({
    email: '',
  });

  const [requestReset, { loading, data, error }] = useMutation(
    REQUEST_RESET_MUTATION,
    {
      variables: inputs,
    }
  );

  async function handleSubmit(e) {
    e.preventDefault();
    // console.log(inputs);
    const res = await requestReset().catch(console.error);
    console.log(res);
    resetForm();
  }

  return (
    <Form method="Post" onSubmit={handleSubmit}>
      <h2>Request Password Reset</h2>
      <DisplayError error={error} />
      <fieldset disabled={loading} aria-busy={loading}>
        {data?.sendUserPasswordResetLink === null && (
          <p>
            If an account exists for that email, check your email for a link!
          </p>
        )}
        <label htmlFor="email">
          Email
          <input
            type="email"
            id="email"
            name="email"
            placeholder="Email"
            autoComplete="email"
            value={inputs.email || ''}
            onChange={handleChange}
          />
        </label>
        <button type="submit">Request Reset</button>
      </fieldset>
    </Form>
  );
}
