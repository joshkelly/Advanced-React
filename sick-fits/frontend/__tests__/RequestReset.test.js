import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import RequestReset, {
  REQUEST_RESET_MUTATION,
} from '../components/RequestReset';

const email = 'test@test.com';

const mocks = [
  {
    request: {
      query: REQUEST_RESET_MUTATION,
      variables: { email },
    },
    result: {
      data: { sendUserPasswordResetLink: null },
    },
  },
];

describe('<RequestReset/>', () => {
  it('renders and matches snapshot', () => {
    const { container, debug } = render(
      <MockedProvider>
        <RequestReset />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('calls the mutation when submitted', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <RequestReset />
      </MockedProvider>
    );
    await userEvent.type(screen.getByPlaceholderText(/Email/i), email);
    await userEvent.click(screen.getByText('Request Reset'));
    await screen.findByText(
      'If an account exists for that email, check your email for a link!'
    );
  });
});
