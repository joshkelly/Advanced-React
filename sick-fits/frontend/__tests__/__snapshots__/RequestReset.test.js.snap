// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset/> renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 hVEEdl"
    method="Post"
  >
    <h2>
      Request Password Reset
    </h2>
    <fieldset
      aria-busy="false"
    >
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          id="email"
          name="email"
          placeholder="Email"
          type="email"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Request Reset
      </button>
    </fieldset>
  </form>
</div>
`;
